### PR TITLE
Feature/hera_preprocess_gefs

### PR DESCRIPTION
**Pull request by hyunchul386**
conv_gefs2datm : Convert input data from GEFS to DATM

### CODE REVIEW

File: utilities/conv_gefs2datm/0Read_me_missing_GEFS [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+2015083018 --> recovered (02/04)
+2016043000 --> recovered (02/04)
+2017031512 --> recovered (02/04)
+2017031618 --> recovered (02/04)
+2017031812 --> recovered (02/04)
+2018041700 --> recovered (01/22)
+2019070806 --> recovered (02/04)
+
+
+#---- massages in error
```
Comment (JessicaMeixner-NOAA) [inline utilities/conv_gefs2datm/0Read_me_missing_GEFS]: Do we have all of the GEFS files now?

File: utilities/conv_gefs2datm/0readme [hunk 2]
```diff
@@ -0,0 +1 @@
+conv_gefs2datm
```
Comment (JessicaMeixner-NOAA) [inline utilities/conv_gefs2datm/0readme]: Can you add instructions on what to do or where to find more information?

File: utilities/conv_gefs2datm/0Read_me_missing_GEFS (hunk 1 above)
Comment (hyunchul386) [reply utilities/conv_gefs2datm/0Read_me_missing_GEFS]: Yes, all missing files in GEFS were recovered by the rerun.

Comment (JessicaMeixner-NOAA) [reply utilities/conv_gefs2datm/0Read_me_missing_GEFS]: Then do we still need this file?

Comment (hyunchul386) [reply utilities/conv_gefs2datm/0Read_me_missing_GEFS]: These files have been processed completely, and saved to HPSS, Hera and DATM_INPUT.

File: utilities/conv_gefs2datm/0readme (hunk 2 above)
Comment (hyunchul386) [reply utilities/conv_gefs2datm/0readme]: some imformation added to 0readme file

Comment (JessicaMeixner-NOAA) [reply utilities/conv_gefs2datm/0readme]: Thanks! This is great information.  Can you also add information on how to run the conversion from the original file format to netcdf? And either list or say where we can find more information on if any assumptions were made or how values were calculated?

File: utilities/conv_gefs2datm/0Read_me_missing_GEFS (hunk 1 above)
Comment (JessicaMeixner-NOAA) [reply utilities/conv_gefs2datm/0Read_me_missing_GEFS]: That's great! I don't think we need this file in the repository then, correct?

**Comment by hyunchul386** [comment]
Correct, we don't need this file at repository.

**Comment by hyunchul386** [comment]
updated 0readme file

File: utilities/conv_gefs2datm/0Read_me_missing_GEFS (hunk 1 above)
Comment (hyunchul386) [reply utilities/conv_gefs2datm/0Read_me_missing_GEFS]: Correct, we dont't need these files in the repository

File: utilities/conv_gefs2datm/0readme (hunk 2 above)
Comment (hyunchul386) [reply utilities/conv_gefs2datm/0readme]: Updated 0readme file

File: utilities/conv_gefs2datm/0Read_me_missing_GEFS (hunk 1 above)
Comment (JessicaMeixner-NOAA) [reply utilities/conv_gefs2datm/0Read_me_missing_GEFS]: Did I miss where you deleted this file then?

File: utilities/conv_gefs2datm/0readme (hunk 2 above)
Comment (JessicaMeixner-NOAA) [reply utilities/conv_gefs2datm/0readme]: Thank you for the instructions on how to run the script.  Can you also put where to find more information about any fields that are modified other than being changed to netcdf format?

**Comment by hyunchul386** [comment]
Updated 0readme file

File: utilities/conv_gefs2datm/0readme (hunk 2 above)
Comment (JessicaMeixner-NOAA) [reply utilities/conv_gefs2datm/0readme]: Thanks!

File: utilities/conv_gefs2datm/conv_gefs2datm.fort.csh [hunk 3]
```diff
@@ -0,0 +1,52 @@
+#! /bin/csh -x
+#-- hyun-chul.lee@noaa.gov
+
+module load NetCDF/4.5.0
```
Comment (flampouris) [inline utilities/conv_gefs2datm/conv_gefs2datm.fort.csh]: This module does not exist on Hera.

In order to load the netcdf module, intel has to be loaded first

`module load intel netcdf`

File: utilities/conv_gefs2datm/conv_gefs2datm.fort.csh [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+set hdir = /scratch2/NCEPDEV/climate/Hyun-Chul.Lee/GEFS
+set sdir = ${hdir}/Reanal
+set tdir = ${hdir}/GEFS2DATM
+set wdir = ${hdir}/Work
+set grb2 = /home/Hyun-Chul.Lee/Tools/wgrib2
```
Comment (flampouris) [inline utilities/conv_gefs2datm/conv_gefs2datm.fort.csh]: This points to a personal home directory with a binary. 

Please update it by loading the appropriate module:

`module load wgrib2`

File: utilities/conv_gefs2datm/conv_gefs2datm.fort.csh [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+#set symd = 20180101
```
Comment (flampouris) [inline utilities/conv_gefs2datm/conv_gefs2datm.fort.csh]: The commented out code should be removed or explain the reason of existence.

File: utilities/conv_gefs2datm/comp_f77_code.csh [hunk 6]
```diff
@@ -0,0 +1,6 @@
+#! /bin/csh
+
+set name = $1
+#--- in hera 
+ifort -132 ${name}.f -o ${name} -L/apps/netcdf/4.6.1/intel/16.1.150/lib -lnetcdff
```
Comment (flampouris) [inline utilities/conv_gefs2datm/comp_f77_code.csh]: the used netcdf library is different (4.6.) than the one loaded at conv_gefs2datm.fort.csh (netcdf 4.5.), please be concistent throughout the app.

**Review by flampouris — CHANGES_REQUESTED** [review]
@hyunchul386 very important job and clean fortran code!

But as is, it does not run for anybody else on Hera.

I look forward to having the code updates to my comments below and merging the code.

File: utilities/conv_gefs2datm/conv_gefs2datm.fort.csh [hunk 7]
```diff
@@ -0,0 +1,52 @@
+#! /bin/csh -x
+#-- hyun-chul.lee@noaa.gov
+
+module load NetCDF/4.5.0
+
+set hdir = /scratch2/NCEPDEV/climate/Hyun-Chul.Lee/GEFS
+set sdir = ${hdir}/Reanal
+set tdir = ${hdir}/GEFS2DATM
+set wdir = ${hdir}/Work
```
Comment (flampouris) [inline utilities/conv_gefs2datm/conv_gefs2datm.fort.csh]: "mkdir -p" for all the *dir is missing

**Comment by hyunchul386** [comment]
Updated conv_gefs2datm.fort.csh

**Review by flampouris — CHANGES_REQUESTED** [review]
@hyunchul386  thank you for the updates. 

Still, the application fails and the conv_gefs2datm/0Read_me_missing_GEFS file is still there. 

Please remove the file, and test the app before the next review iteration.

Thank you.

**Comment by hyunchul386** [comment]
I have used it, and let me know what you have failed.

**Comment by flampouris** [comment]
From my original review:
> module load **intel** netcdf

The scripts are using intel compiler without loading the appropriate modules.

Please, check out your branch to a different directory and test it. The application needs a couple more updates to stop failing.

**Comment by hyunchul386** [comment]
Updated comp_f77_code.csh.

File: utilities/conv_gefs2datm/comp_f77_code.csh [hunk 8]
```diff
@@ -0,0 +1,8 @@
+#! /bin/csh
+module load intel/19.0.4.243
```
Comment (flampouris) [inline utilities/conv_gefs2datm/comp_f77_code.csh]: intel19 is buggy, lets use the default: 

module load intel

File: utilities/conv_gefs2datm/comp_f77_code.csh [hunk 9]
```diff
@@ -0,0 +1,8 @@
...
+module load impi/2019.0.4
```
Comment (flampouris) [inline utilities/conv_gefs2datm/comp_f77_code.csh]: impi is not required for this code

File: utilities/conv_gefs2datm/comp_f77_code.csh [hunk 10]
```diff
@@ -0,0 +1,8 @@
...
+
```
Comment (flampouris) [inline utilities/conv_gefs2datm/comp_f77_code.csh:4]: `module load netcdf` to be added

File: utilities/conv_gefs2datm/comp_f77_code.csh [hunk 11]
```diff
@@ -0,0 +1,8 @@
...
+set name = $1
+#--- in hera
```
Comment (flampouris) [inline utilities/conv_gefs2datm/comp_f77_code.csh]: Clean code --> Easier to make a Makefile:

set FFLAGS='extend-source 132'

File: utilities/conv_gefs2datm/comp_f77_code.csh [hunk 12]
```diff
@@ -0,0 +1,8 @@
+#! /bin/csh
+module load intel/19.0.4.243
+module load impi/2019.0.4
+
+set name = $1
+#--- in hera 
+ifort -132 ${name}.f -o ${name} -L/apps/netcdf/4.6.1/intel/16.1.150/lib -lnetcdff
```
Comment (flampouris) [inline utilities/conv_gefs2datm/comp_f77_code.csh]: 1. There is no `include` in this line, it's impossible to compile this code!!!

2. updated compilation 
ifort $FFLAGS ${name}.f -o ${name} -I$NETCDF/include -L$NETCDF/lib -lnetcdff

File: utilities/conv_gefs2datm/comp_f77_code.csh [hunk 13]
```diff
@@ -0,0 +1,8 @@
...
+
```
Comment (flampouris) [inline utilities/conv_gefs2datm/comp_f77_code.csh:8]: Remove blank lines

File: utilities/conv_gefs2datm/conv_gefs2datm.fort.csh [hunk 14]
```diff
@@ -0,0 +1,35 @@
+#! /bin/csh -x
+
+module load netcdf/4.6.1
```
Comment (flampouris) [inline utilities/conv_gefs2datm/conv_gefs2datm.fort.csh]: intel has to be loaded before netcdf

module load intel
module load netcdf

File: utilities/conv_gefs2datm/conv_gefs2datm.f [hunk 15]
```diff
@@ -0,0 +1,888 @@
...
+      character*(*) FILE_NAME, FOUT_NAME
+      character STR*200,SATT*200
+      parameter (FILE_NAME = "./gefs_input.nc")
+      parameter (FOUT_NAME = "./gefs_output.nc")
+
+C C We are reading 
+      integer NX,NY,NT,nerr,NDIM,i,j,k,NSTR
+      parameter (NX = 1536, NY = 768, NT = 1, NDIM = 3)
+      real*4 :: latitude(NY), longitude(NX)
+      integer time(NT),NF_64BIT_OFFSET,vi_indx
```
Comment (flampouris) [inline utilities/conv_gefs2datm/conv_gefs2datm.f]: The NF_64BIT_OFFSET is not used in the code and probably defined at the netcdf library, it has to be removed from here.

**Review by flampouris — CHANGES_REQUESTED** [review]
@hyunchul386 please go through the last bunch of my comments and modify your code and scripts.

**Comment by hyunchul386** [comment]
No netcdf is used here

**Comment by hyunchul386** [comment]
I think it is okay with blank line

**Comment by hyunchul386** [comment]
Updated.
Just curious, is intel 19 buggy for f77?

**Comment by flampouris** [comment]
> No netcdf is used here

it is used to compile the code...

**Comment by hyunchul386** [comment]
Actually, " module load netcdf/4.6.1  " in conv_gefs2datm.fort.csh

**Review by flampouris — APPROVED** [review]

**Review by JessicaMeixner-NOAA — APPROVED** [review]

**Pull request merged**